From c330f345162f90520955fd44268b8e140394d9a2 Mon Sep 17 00:00:00 2001
From: leoliu-oc <leoliu-oc@zhaoxin.com>
Date: Fri, 14 Jun 2024 11:10:08 +0800
Subject: [PATCH 175/360] pinctrl: Add support Zhaoxin GPIO pinctrl

zhaoxin inclusion
category: feature

--------------------

Implements gpio interrupt and gpio management functions and provides
standard pinctrl and gpio interfaces.

Signed-off-by: leoliu-oc <leoliu-oc@zhaoxin.com>
---
 drivers/pinctrl/Kconfig                   |   1 +
 drivers/pinctrl/Makefile                  |   1 +
 drivers/pinctrl/zhaoxin/Kconfig           |  22 +
 drivers/pinctrl/zhaoxin/Makefile          |   4 +
 drivers/pinctrl/zhaoxin/pinctrl-kx7000.c  | 354 ++++++++++
 drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.c | 756 ++++++++++++++++++++++
 drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.h | 135 ++++
 7 files changed, 1273 insertions(+)
 create mode 100644 drivers/pinctrl/zhaoxin/Kconfig
 create mode 100644 drivers/pinctrl/zhaoxin/Makefile
 create mode 100644 drivers/pinctrl/zhaoxin/pinctrl-kx7000.c
 create mode 100644 drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.c
 create mode 100644 drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.h

diff --git a/drivers/pinctrl/Kconfig b/drivers/pinctrl/Kconfig
index 354536de564b..86f3cfc03fe5 100644
--- a/drivers/pinctrl/Kconfig
+++ b/drivers/pinctrl/Kconfig
@@ -594,6 +594,7 @@ source "drivers/pinctrl/berlin/Kconfig"
 source "drivers/pinctrl/cirrus/Kconfig"
 source "drivers/pinctrl/freescale/Kconfig"
 source "drivers/pinctrl/intel/Kconfig"
+source "drivers/pinctrl/zhaoxin/Kconfig"
 source "drivers/pinctrl/mediatek/Kconfig"
 source "drivers/pinctrl/meson/Kconfig"
 source "drivers/pinctrl/mvebu/Kconfig"
diff --git a/drivers/pinctrl/Makefile b/drivers/pinctrl/Makefile
index 97823f52b972..59b6d8a8d7f9 100644
--- a/drivers/pinctrl/Makefile
+++ b/drivers/pinctrl/Makefile
@@ -64,6 +64,7 @@ obj-$(CONFIG_PINCTRL_BERLIN)	+= berlin/
 obj-y				+= cirrus/
 obj-y				+= freescale/
 obj-$(CONFIG_X86)		+= intel/
+obj-$(CONFIG_X86)		+= zhaoxin/
 obj-y				+= mediatek/
 obj-$(CONFIG_PINCTRL_MESON)	+= meson/
 obj-y				+= mvebu/
diff --git a/drivers/pinctrl/zhaoxin/Kconfig b/drivers/pinctrl/zhaoxin/Kconfig
new file mode 100644
index 000000000000..dece29a29423
--- /dev/null
+++ b/drivers/pinctrl/zhaoxin/Kconfig
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: GPL-2.0
+# Intel pin control drivers
+
+if (X86 || COMPILE_TEST)
+
+config PINCTRL_ZHAOXIN
+	tristate
+	select PINMUX
+	select PINCONF
+	select GENERIC_PINCONF
+	select GPIOLIB
+	select GPIOLIB_IRQCHIP
+
+config PINCTRL_KX7000
+	tristate "Zhaoxin KX7000 pinctrl and GPIO driver"
+	depends on ACPI
+	select PINCTRL_ZHAOXIN
+	help
+	  This pinctrl driver provides an interface that allows configuring
+	  of Zhaoxin KX7000 chipset pins and using them as GPIOs.
+
+endif
diff --git a/drivers/pinctrl/zhaoxin/Makefile b/drivers/pinctrl/zhaoxin/Makefile
new file mode 100644
index 000000000000..a3acfa66f196
--- /dev/null
+++ b/drivers/pinctrl/zhaoxin/Makefile
@@ -0,0 +1,4 @@
+# zhaoxin pin control drivers
+
+obj-$(CONFIG_PINCTRL_ZHAOXIN)		+= pinctrl-zhaoxin.o
+obj-$(CONFIG_PINCTRL_KX7000)		+= pinctrl-kx7000.o
diff --git a/drivers/pinctrl/zhaoxin/pinctrl-kx7000.c b/drivers/pinctrl/zhaoxin/pinctrl-kx7000.c
new file mode 100644
index 000000000000..8e83df0ef2a7
--- /dev/null
+++ b/drivers/pinctrl/zhaoxin/pinctrl-kx7000.c
@@ -0,0 +1,354 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * zhaoxin KX7000 pinctrl/GPIO driver
+ *
+ * Copyright(c) 2023 Shanghai Zhaoxin Corporation. All rights reserved.
+ *
+ */
+
+#define DRIVER_VERSION "1.0.0"
+
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+
+#include <linux/pinctrl/pinctrl.h>
+
+#include "pinctrl-zhaoxin.h"
+
+#define ZX_CAL_ARRAY(a, b) \
+{ \
+	.pmio_offset = (a), \
+	.size = (b), \
+}
+
+#define PMIO_RX90 100
+#define PMIO_RX8C 200
+
+#define ZX_CAL_INDEX_ARRAY(a, b, c) \
+{ \
+	.reg_port_base = (PMIO_RX90), \
+	.reg_data_base = (PMIO_RX8C), \
+	.index = (a), \
+	.cal_array = (b), \
+	.size = (c), \
+}
+
+/* kx7000 pin define */
+static const struct pinctrl_pin_desc kx7000_pins[] = {
+
+	PINCTRL_PIN(0, "IOD_CPUTCK"),
+	PINCTRL_PIN(1, "IOD_CPUTMS"),
+	PINCTRL_PIN(2, "IOD_CPUTRST"),
+	PINCTRL_PIN(3, "IOD_CPUTDO"),
+	PINCTRL_PIN(4, "IOD_CPUTDI"),
+	PINCTRL_PIN(5, "IOD_ZLSCLK0"),
+	PINCTRL_PIN(6, "IOD_ZLDATA0"),
+	PINCTRL_PIN(7, "IOD_ZLSCLK1"),
+	PINCTRL_PIN(8, "IOD_ZLDATA1"),
+	PINCTRL_PIN(9, "IOD_CLK27M"),
+	PINCTRL_PIN(10, "IOD_CPURST"),
+	PINCTRL_PIN(11, "IOD_PWORK"),
+	PINCTRL_PIN(12, "IOD_RSMRST"),
+	PINCTRL_PIN(13, "IOD_THRMTRIP"),
+	//GPIO range 0
+	PINCTRL_PIN(14, "USBHOC0"),
+	PINCTRL_PIN(15, "USBHOC1"),
+	PINCTRL_PIN(16, "USBHOC2"),
+	PINCTRL_PIN(17, "USBHOC3"),
+	PINCTRL_PIN(18, "USBHOC4"),
+	PINCTRL_PIN(19, "USBHOC5"),
+	PINCTRL_PIN(20, "USBHOC6"),
+	PINCTRL_PIN(21, "USBHOC7"),
+	//gpio range 1
+	PINCTRL_PIN(22, "USB4SBTX0"),
+	PINCTRL_PIN(23, "USB4SBRX0"),
+	PINCTRL_PIN(24, "USB4SBTX1"),
+	PINCTRL_PIN(25, "USB4SBRX1"),
+	//gpio range 2
+	PINCTRL_PIN(26, "I2C1DT"),
+	PINCTRL_PIN(27, "I2C1CK"),
+	PINCTRL_PIN(28, "I2C1INT"),
+	//gpio range 3
+	PINCTRL_PIN(29, "I2C2DT"),
+	PINCTRL_PIN(30, "I2C2CK"),
+	//gpio range 4
+	PINCTRL_PIN(31, "I2C2INT"),
+	//gpio range 5
+	PINCTRL_PIN(32, "SMBDT1"),
+	PINCTRL_PIN(33, "SMBCK1"),
+	PINCTRL_PIN(34, "SMBDT2"),
+	PINCTRL_PIN(35, "SMBCK2"),
+	PINCTRL_PIN(36, "SMBALRT"),
+	//gpio range 6
+	PINCTRL_PIN(37, "SME_I2CDT"),
+	PINCTRL_PIN(38, "SME_I2CCK"),
+	//gpio range 7
+	PINCTRL_PIN(39, "PWM"),
+	PINCTRL_PIN(40, "TACH"),
+	//gpio range 8
+	PINCTRL_PIN(41, "GPIO0"),
+	PINCTRL_PIN(42, "GPIO1"),
+	PINCTRL_PIN(43, "GPIO2"),
+	PINCTRL_PIN(44, "GPIO3"),
+	PINCTRL_PIN(45, "GPIO4"),
+	PINCTRL_PIN(46, "GPIO5"),
+	PINCTRL_PIN(47, "GPIO6"),
+	PINCTRL_PIN(48, "GPIO7"),
+	PINCTRL_PIN(49, "GPIO8"),
+	PINCTRL_PIN(50, "GPIO9"),
+	PINCTRL_PIN(51, "LPCCLK"),
+	PINCTRL_PIN(52, "LPCDRQ1"),
+	//gpio range 9
+	PINCTRL_PIN(53, "LPCDRQ0"),
+	PINCTRL_PIN(54, "LPCFRAME"),
+	PINCTRL_PIN(55, "LPCAD3"),
+	PINCTRL_PIN(56, "LPCAD2"),
+	PINCTRL_PIN(57, "LPCAD1"),
+	PINCTRL_PIN(58, "LPCAD0"),
+	//gpio range 10
+	PINCTRL_PIN(59, "SERIRQ"),
+	PINCTRL_PIN(60, "AZRST"),
+	PINCTRL_PIN(61, "AZBITCLK"),
+	PINCTRL_PIN(62, "AZSDIN0"),
+	PINCTRL_PIN(63, "AZSDIN1"),
+	PINCTRL_PIN(64, "AZSDOUT"),
+	PINCTRL_PIN(65, "AZSYNC"),
+	//gpio range 11
+	PINCTRL_PIN(66, "I2S1_SCLK"),
+	PINCTRL_PIN(67, "I2S1_TXD"),
+	PINCTRL_PIN(68, "I2S1_WS"),
+	PINCTRL_PIN(69, "I2S1_MCLK"),
+	//gpio range 12
+	PINCTRL_PIN(70, "I2S1_RXD"),
+	//gpio range 13
+	PINCTRL_PIN(71, "I2S1_INT"),
+	PINCTRL_PIN(72, "MSPIDI"),
+	PINCTRL_PIN(73, "MSPIDO"),
+	PINCTRL_PIN(74, "MSPIIO2"),
+	PINCTRL_PIN(75, "MSPIIO3"),
+	PINCTRL_PIN(76, "MSPICLK"),
+	PINCTRL_PIN(77, "MSPISS0"),
+	//gpio range 14
+	PINCTRL_PIN(78, "MSPISS1"),
+	PINCTRL_PIN(79, "MSPISS2"),
+	//gpio range 15
+	PINCTRL_PIN(80, "SPIDEVINT"),
+	PINCTRL_PIN(81, "BIOSSEL"),
+	//gpio range 16
+	PINCTRL_PIN(82, "THRM"),
+	PINCTRL_PIN(83, "PEXWAKE"),
+	PINCTRL_PIN(84, "PWRBTN"),
+	//gpio range 17
+	PINCTRL_PIN(85, "SPKR"),
+	PINCTRL_PIN(86, "PME"),
+	//gpio range 18
+	PINCTRL_PIN(87, "BATLOW"),
+	PINCTRL_PIN(88, "EXTSMI"),
+	PINCTRL_PIN(89, "SUSA"),
+	PINCTRL_PIN(90, "SUSB"),
+	PINCTRL_PIN(91, "SUSC"),
+	PINCTRL_PIN(92, "GPWAKE"),
+	PINCTRL_PIN(93, "RING"),
+	PINCTRL_PIN(94, "LID"),
+	PINCTRL_PIN(95, "SLPS0"),
+	PINCTRL_PIN(96, "PCIRST"),
+	PINCTRL_PIN(97, "SVID_VREN"),
+	//gpio range 19
+	PINCTRL_PIN(98, "INTRUDER"),
+	//gpio range 20
+	PINCTRL_PIN(99, "GFX_I2CCLK0"),
+	PINCTRL_PIN(100, "GFX_I2CDAT0"),
+	PINCTRL_PIN(101, "GFX_I2CCLK1"),
+	PINCTRL_PIN(102, "GFX_I2CDAT1"),
+	PINCTRL_PIN(103, "GFX_I2CCLK2"),
+	PINCTRL_PIN(104, "GFX_I2CDAT2"),
+	PINCTRL_PIN(105, "GFX_I2CCLK3"),
+	PINCTRL_PIN(106, "GFX_I2CDAT3"),
+	PINCTRL_PIN(107, "GFX_GPIO0"),
+	PINCTRL_PIN(108, "GFX_GPIO1"),
+	PINCTRL_PIN(109, "GFX_GPIO2"),
+	PINCTRL_PIN(110, "GFX_GPIO3"),
+	PINCTRL_PIN(111, "CRTHSYNC"),
+	PINCTRL_PIN(112, "CRTVSYNC"),
+};
+
+#define NOT_DEFINE	-30000
+
+static int calibrate_int[] = {
+	0, 1, 2, 3, 4, 5, 6, 7, 8, 9,
+	63, 64, 65, 66, 67, 68,
+	69, 70,
+	18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
+	34, 35, 36, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62
+};
+
+static int calibrate_sattus[] = {
+	0, 1, 2, 3, 4, 5, 6, 7, 8, 9,
+	63, 64, 65, 66, 67, 68,
+	69, 70,
+	18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
+	34, 35, 36, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62
+};
+
+static const struct reg_cal_array kx7000_int_cal[] = {
+	ZX_CAL_ARRAY(0x58, 16),
+	ZX_CAL_ARRAY(0x5A, 2),
+	ZX_CAL_ARRAY(0xDA, 16),
+	ZX_CAL_ARRAY(0xDE, 16),
+};
+
+static const struct reg_calibrate int_cal[] = {
+	{
+		.reg = kx7000_int_cal,
+		.reg_cal_size = ARRAY_SIZE(kx7000_int_cal),
+		.cal_array = calibrate_int,
+		.size = ARRAY_SIZE(calibrate_int),
+	}
+};
+
+static const struct reg_cal_array kx7000_status_cal[] = {
+	ZX_CAL_ARRAY((0x8), 16),
+	ZX_CAL_ARRAY((0xE), 2),
+	ZX_CAL_ARRAY((0xA), 16),
+	ZX_CAL_ARRAY((0xC), 16),
+};
+
+static const struct reg_calibrate status_cal[] = {
+	{
+		.reg = kx7000_status_cal,
+		.reg_cal_size = ARRAY_SIZE(kx7000_status_cal),
+		.cal_array = calibrate_sattus,
+		.size = ARRAY_SIZE(calibrate_sattus),
+	}
+};
+
+static const struct reg_cal_array kx7000_mod_sel_cal[] = {
+	ZX_CAL_ARRAY((0x0), 16),
+	ZX_CAL_ARRAY((0x6), 2),
+	ZX_CAL_ARRAY((0x2), 16),
+	ZX_CAL_ARRAY((0x4), 16),
+};
+
+static const struct reg_calibrate mod_sel_cal[] = {
+	{
+		.reg = kx7000_mod_sel_cal,
+		.reg_cal_size = ARRAY_SIZE(kx7000_mod_sel_cal),
+		.cal_array = calibrate_sattus,
+		.size = ARRAY_SIZE(calibrate_sattus),
+	}
+};
+
+static const struct index_cal_array kx7000_gpio_in_cal[] = {
+	ZX_CAL_INDEX_ARRAY(0x98, NULL, 71),
+};
+
+static const struct index_cal_array kx7000_gpio_out_cal[] = {
+	ZX_CAL_INDEX_ARRAY(0x90, NULL, 71),
+};
+
+static int calibrate_trigger[] = {
+	0, 1, 2, 3, 4, 5, 6, 7,
+	8, 9, 18, 19,
+	20, 21, 22, 23,
+	24, 25, 26, 27,
+	28, 29, 30, 31,
+	32, 33, 34, 35,
+	36, 50, 51, 52,
+	53, 54, 55, 56,
+	57, 58, 59, 60,
+	61, 62, 63, 64,
+	65, 66, 67, 68,
+	69, 70
+};
+
+static const struct index_cal_array kx7000_trigger_cal[] = {
+	ZX_CAL_INDEX_ARRAY(0xA0, calibrate_trigger, 50),
+};
+
+static const struct zhaoxin_pin_topology kx7000_pin_topologys[] = {
+	{
+		.int_cal = int_cal,
+		.status_cal = status_cal,
+		.mod_sel_cal = mod_sel_cal,
+		.gpio_in_cal = kx7000_gpio_in_cal,
+		.gpio_out_cal = kx7000_gpio_out_cal,
+		.trigger_cal = kx7000_trigger_cal,
+	}
+};
+
+#define KX7000_GPP(s, e, g) \
+{ \
+	.zhaoxin_range_pin_base = (s), \
+	.zhaoxin_range_pin_size = ((e) - (s) + 1), \
+	.zhaoxin_range_gpio_base = (g), \
+}
+
+static const struct zhaoxin_pin_map2_gpio kx7000_pinmap_gpps[] = {
+	KX7000_GPP(0, 13, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(14, 19, 10),
+	KX7000_GPP(20, 21, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(22, 25, 65),
+	KX7000_GPP(26, 28, 43),
+	KX7000_GPP(29, 30, 41),
+	KX7000_GPP(31, 31, 49),
+	KX7000_GPP(32, 36, 16),
+	KX7000_GPP(37, 38, 69),
+	KX7000_GPP(39, 40, 67),
+	KX7000_GPP(41, 50, 0),
+	KX7000_GPP(51, 52, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(53, 53, 39),
+	KX7000_GPP(54, 58, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(59, 59, 40),
+	KX7000_GPP(60, 65, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(66, 69, 35),
+	KX7000_GPP(70, 70, 46),
+	KX7000_GPP(71, 71, 64),
+	KX7000_GPP(72, 77, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(78, 78, 50),
+	KX7000_GPP(79, 79, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(80, 80, 51),
+	KX7000_GPP(81, 81, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(82, 82, 52),
+	KX7000_GPP(83, 84, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(85, 85, 53),
+	KX7000_GPP(86, 86, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(87, 95, 54),
+	KX7000_GPP(96, 97, ZHAOXIN_GPIO_BASE_NOMAP),
+	KX7000_GPP(98, 98, 63),
+	KX7000_GPP(99, 112, 21),
+};
+
+static const struct zhaoxin_pinctrl_soc_data kx7000_soc_data = {
+	.pins = kx7000_pins,
+	.npins = ARRAY_SIZE(kx7000_pins),
+	.pin_topologys = kx7000_pin_topologys,
+	.zhaoxin_pin_maps = kx7000_pinmap_gpps,
+	.pin_map_size = ARRAY_SIZE(kx7000_pinmap_gpps),
+};
+
+static const struct acpi_device_id kx7000_pinctrl_acpi_match[] = {
+	{ "KX8344B", (kernel_ulong_t)&kx7000_soc_data },
+	{ }
+};
+MODULE_DEVICE_TABLE(acpi, kx7000_pinctrl_acpi_match);
+
+static const struct dev_pm_ops kx7000_pinctrl_pm_ops = {
+	SET_NOIRQ_SYSTEM_SLEEP_PM_OPS(zhaoxin_pinctrl_suspend_noirq, zhaoxin_pinctrl_resume_noirq)
+};
+
+static struct platform_driver kx7000_pinctrl_driver = {
+	.probe = zhaoxin_pinctrl_probe_by_hid,
+	.driver = {
+		.name = "kx7000-pinctrl",
+		.acpi_match_table = kx7000_pinctrl_acpi_match,
+		.pm = &kx7000_pinctrl_pm_ops,
+	},
+};
+
+module_platform_driver(kx7000_pinctrl_driver);
+
+MODULE_AUTHOR("www.zhaoxin.com");
+MODULE_DESCRIPTION("Shanghai Zhaoxin pinctrl driver");
+MODULE_VERSION(DRIVER_VERSION);
+MODULE_LICENSE("GPL v2");
diff --git a/drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.c b/drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.c
new file mode 100644
index 000000000000..b639f2cd079c
--- /dev/null
+++ b/drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.c
@@ -0,0 +1,756 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ *    zhaoxin pinctrl common code
+ *
+ *    Copyright(c) 2021 Shanghai Zhaoxin Corporation. All rights reserved.
+ *
+ */
+
+#define DRIVER_VERSION "1.0.0"
+
+#include <linux/acpi.h>
+#include <linux/gpio/driver.h>
+#include <linux/interrupt.h>
+#include <linux/log2.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/property.h>
+#include <linux/time.h>
+
+#include <linux/pinctrl/pinctrl.h>
+#include <linux/pinctrl/pinmux.h>
+#include <linux/pinctrl/pinconf.h>
+#include <linux/pinctrl/pinconf-generic.h>
+
+#include "../core.h"
+#include "pinctrl-zhaoxin.h"
+
+static int pin_to_hwgpio(struct pinctrl_gpio_range *range, unsigned int pin)
+{
+	int offset = 0;
+
+	if (range->pins) {
+		for (offset = 0; offset < range->npins; offset++)
+			if (pin == range->pins[offset])
+				break;
+		return range->base+offset-range->gc->base;
+	} else
+		return pin-range->pin_base+range->base-range->gc->base;
+}
+
+static u16 zx_pad_read16(struct zhaoxin_pinctrl *pctrl, u8 index)
+{
+	outb(index, pctrl->pmio_rx90+pctrl->pmio_base);
+	return inw(pctrl->pmio_rx8c+pctrl->pmio_base);
+}
+
+static void zx_pad_write16(struct zhaoxin_pinctrl *pctrl, u8 index, u16 value)
+{
+	outb(index, pctrl->pmio_rx90+pctrl->pmio_base);
+	outw(value, pctrl->pmio_rx8c+pctrl->pmio_base);
+}
+
+static int zhaoxin_get_groups_count(struct pinctrl_dev *pctldev)
+{
+	struct zhaoxin_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
+
+	return pctrl->soc->ngroups;
+}
+
+static const char *zhaoxin_get_group_name(struct pinctrl_dev *pctldev, unsigned int group)
+{
+	struct zhaoxin_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
+
+	return pctrl->soc->groups[group].name;
+}
+
+static int zhaoxin_get_group_pins(struct pinctrl_dev *pctldev, unsigned int group,
+				  const unsigned int **pins, unsigned int *npins)
+{
+	struct zhaoxin_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
+
+	*pins = pctrl->soc->groups[group].pins;
+	*npins = pctrl->soc->groups[group].npins;
+
+	return 0;
+}
+
+static void zhaoxin_pin_dbg_show(struct pinctrl_dev *pctldev, struct seq_file *s, unsigned int pin)
+{
+
+}
+
+static const struct pinctrl_ops zhaoxin_pinctrl_ops = {
+	.get_groups_count = zhaoxin_get_groups_count,
+	.get_group_name = zhaoxin_get_group_name,
+	.get_group_pins = zhaoxin_get_group_pins,
+	.pin_dbg_show = zhaoxin_pin_dbg_show,
+};
+
+static int zhaoxin_get_functions_count(struct pinctrl_dev *pctldev)
+{
+	struct zhaoxin_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
+
+	return pctrl->soc->nfunctions;
+}
+
+static const char *zhaoxin_get_function_name(struct pinctrl_dev *pctldev, unsigned int function)
+{
+	struct zhaoxin_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
+
+	return pctrl->soc->functions[function].name;
+}
+
+static int zhaoxin_get_function_groups(struct pinctrl_dev *pctldev, unsigned int function,
+				       const char *const **groups, unsigned int *const ngroups)
+{
+	struct zhaoxin_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
+
+	*groups = pctrl->soc->functions[function].groups;
+	*ngroups = pctrl->soc->functions[function].ngroups;
+
+	return 0;
+}
+
+static int zhaoxin_pinmux_set_mux(struct pinctrl_dev *pctldev, unsigned int function,
+				  unsigned int group)
+{
+	return 0;
+}
+
+#define ZHAOXIN_PULL_UP_20K		0x80
+#define ZHAOXIN_PULL_UP_10K		0x40
+#define ZHAOXIN_PULL_UP_47K		0x20
+#define ZHAOXIN_PULL_DOWN		0x10
+
+#define ZHAOXIN_PULL_UP	0xe0
+
+static void zhaoxin_gpio_set_gpio_mode_and_pull(struct zhaoxin_pinctrl *pctrl, unsigned int pin,
+						bool isup)
+{
+	u16 tmp = 0;
+	u16 value;
+	u16 value_back = 0;
+
+	if (isup)
+		tmp = ZHAOXIN_PULL_UP_10K|1;
+	else
+		tmp = ZHAOXIN_PULL_DOWN|1;
+	value = zx_pad_read16(pctrl, pin);
+
+	if (pin <= 0x32 && pin >= 0x29) {
+		if (isup) {
+			value &= (~(ZHAOXIN_PULL_DOWN));
+			value |= tmp;
+		} else {
+			value &= (~(ZHAOXIN_PULL_UP));
+			value |= tmp;
+		}
+		value &= ~(0x1);
+		zx_pad_write16(pctrl, pin, value);
+		value_back = zx_pad_read16(pctrl, pin);
+	} else {
+		if (isup) {
+			value &= (~(ZHAOXIN_PULL_DOWN));
+			value |= tmp;
+		} else {
+			value &= (~(ZHAOXIN_PULL_UP));
+			value |= tmp;
+		}
+		value |= 0x1;
+		zx_pad_write16(pctrl, pin, value);
+		value_back = zx_pad_read16(pctrl, pin);
+	}
+}
+
+static int zhaoxin_gpio_request_enable(struct pinctrl_dev *pctldev,
+				       struct pinctrl_gpio_range *range, unsigned int pin)
+{
+	struct zhaoxin_pinctrl *pctrl = pinctrl_dev_get_drvdata(pctldev);
+
+	zhaoxin_gpio_set_gpio_mode_and_pull(pctrl, pin, true);
+
+	return 0;
+}
+
+static const struct pinmux_ops zhaoxin_pinmux_ops = {
+	.get_functions_count = zhaoxin_get_functions_count,
+	.get_function_name = zhaoxin_get_function_name,
+	.get_function_groups = zhaoxin_get_function_groups,
+	.set_mux = zhaoxin_pinmux_set_mux,
+	.gpio_request_enable = zhaoxin_gpio_request_enable,
+};
+
+static int zhaoxin_config_get(struct pinctrl_dev *pctldev, unsigned int pin, unsigned long *config)
+{
+	return 0;
+}
+
+static int zhaoxin_config_set(struct pinctrl_dev *pctldev, unsigned int pin, unsigned long *configs,
+			      unsigned int nconfigs)
+{
+	return 0;
+}
+
+static const struct pinconf_ops zhaoxin_pinconf_ops = {
+	.is_generic = true,
+	.pin_config_get = zhaoxin_config_get,
+	.pin_config_set = zhaoxin_config_set,
+};
+
+static const struct pinctrl_desc zhaoxin_pinctrl_desc = {
+	.pctlops = &zhaoxin_pinctrl_ops,
+	.pmxops = &zhaoxin_pinmux_ops,
+	.confops = &zhaoxin_pinconf_ops,
+	.owner = THIS_MODULE,
+};
+
+static int zhaoxin_gpio_to_pin(struct zhaoxin_pinctrl *pctrl, unsigned int offset,
+			       const struct zhaoxin_pin_topology **community,
+			       const struct zhaoxin_pin_map2_gpio **padgrp)
+{
+	int i;
+
+	for (i = 0; i < pctrl->pin_map_size; i++) {
+		const struct zhaoxin_pin_map2_gpio *map = &pctrl->pin_maps[i];
+
+		if (map->zhaoxin_range_gpio_base == ZHAOXIN_GPIO_BASE_NOMAP)
+			continue;
+		if (offset >= map->zhaoxin_range_gpio_base &&
+		    offset < map->zhaoxin_range_gpio_base + map->zhaoxin_range_pin_size) {
+			int pin;
+
+			pin = map->zhaoxin_range_pin_base + offset - map->zhaoxin_range_gpio_base;
+			if (padgrp)
+				*padgrp = map;
+			return pin;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static __maybe_unused int zhaoxin_pin_to_gpio(struct zhaoxin_pinctrl *pctrl, int pin)
+{
+	const struct zhaoxin_pin_map2_gpio *pin_maps;
+
+	pin_maps = pctrl->pin_maps;
+	if (!pin_maps)
+		return -EINVAL;
+
+	return pin - pin_maps->zhaoxin_range_pin_base + pin_maps->zhaoxin_range_gpio_base;
+}
+
+static int zhaoxin_gpio_get(struct gpio_chip *chip, unsigned int offset)
+{
+	struct zhaoxin_pinctrl *pctrl = gpiochip_get_data(chip);
+	const struct index_cal_array *gpio_in_cal;
+	int gap = offset/16;
+	int bit = offset%16;
+	int pin;
+	int value;
+
+	gpio_in_cal = pctrl->pin_topologys->gpio_in_cal;
+	pin = zhaoxin_gpio_to_pin(pctrl, offset, NULL, NULL);
+	value = zx_pad_read16(pctrl, gpio_in_cal->index+gap);
+	value &= (1<<bit);
+
+	return !!value;
+}
+
+static void zhaoxin_gpio_set(struct gpio_chip *chip, unsigned int offset, int value)
+{
+	struct zhaoxin_pinctrl *pctrl = gpiochip_get_data(chip);
+	const struct index_cal_array *gpio_out_cal;
+	unsigned long flags;
+	int gap = offset / 16;
+	int bit = offset % 16;
+	u16 org;
+	int pin;
+
+	gpio_out_cal = pctrl->pin_topologys->gpio_out_cal;
+	pin = zhaoxin_gpio_to_pin(pctrl, offset, NULL, NULL);
+
+	raw_spin_lock_irqsave(&pctrl->lock, flags);
+
+	org = zx_pad_read16(pctrl, gpio_out_cal->index+gap);
+	if (value)
+		org |= (1<<bit);
+	else
+		org &= (~(1<<bit));
+	zx_pad_write16(pctrl, gpio_out_cal->index+gap, org);
+	raw_spin_unlock_irqrestore(&pctrl->lock, flags);
+}
+
+static int zhaoxin_gpio_direction_input(struct gpio_chip *chip, unsigned int offset)
+{
+	return pinctrl_gpio_direction_input(chip->base + offset);
+}
+
+static int zhaoxin_gpio_direction_output(struct gpio_chip *chip, unsigned int offset, int value)
+{
+	return pinctrl_gpio_direction_output(chip->base + offset);
+}
+
+static int zhaoxin_gpio_request(struct gpio_chip *gc, unsigned int offset)
+{
+	return gpiochip_generic_request(gc, offset);
+}
+static void zhaoxin_gpio_free(struct gpio_chip *gc, unsigned int offset)
+{
+	gpiochip_generic_free(gc, offset);
+}
+
+static int zhaoxin_gpio_config(struct gpio_chip *gc, unsigned int offset, unsigned long config)
+{
+	return gpiochip_generic_config(gc, offset, config);
+}
+
+static const struct gpio_chip zhaoxin_gpio_chip = {
+	.owner = THIS_MODULE,
+	.request = zhaoxin_gpio_request,
+	.free = zhaoxin_gpio_free,
+	.direction_input = zhaoxin_gpio_direction_input,
+	.direction_output = zhaoxin_gpio_direction_output,
+	.get = zhaoxin_gpio_get,
+	.set = zhaoxin_gpio_set,
+	.set_config = zhaoxin_gpio_config,
+};
+
+static void zhaoxin_gpio_irq_ack(struct irq_data *d)
+{
+	struct gpio_chip *gc = irq_data_get_irq_chip_data(d);
+	struct zhaoxin_pinctrl *pctrl = gpiochip_get_data(gc);
+	const struct reg_calibrate *status_cal;
+	const struct reg_cal_array *reg_off;
+	int gpio = irqd_to_hwirq(d);
+	int i, j;
+	int offset = 0;
+	int base_offset = 0;
+	int bit_off = 0;
+	u16 value;
+	u16 value_read;
+
+	status_cal = pctrl->pin_topologys->status_cal;
+	if (gpio >= 0) {
+		for (i = 0; i < status_cal->size; i++)
+			if (gpio == status_cal->cal_array[i])
+				break;
+		for (j = 0; j < status_cal->reg_cal_size; j++) {
+			if (offset > i)
+				break;
+			offset += status_cal->reg[j].size;
+		}
+		reg_off = &status_cal->reg[j-1];
+		bit_off = i-(offset-reg_off->size);
+		base_offset = reg_off->pmio_offset;
+		value = readw(pctrl->pm_pmio_base+reg_off->pmio_offset);
+		value_read = value;
+		value |= (1<<bit_off);
+		writew(value, pctrl->pm_pmio_base+reg_off->pmio_offset);
+	}
+}
+
+static void zhaoxin_gpio_irq_mask_unmask(struct irq_data *d, bool mask)
+{
+	struct gpio_chip *gc = irq_data_get_irq_chip_data(d);
+	struct zhaoxin_pinctrl *pctrl = gpiochip_get_data(gc);
+	const struct reg_calibrate *int_cal;
+	const struct reg_calibrate *mod_sel_cal;
+	int gpio = irqd_to_hwirq(d);
+	int i, j;
+	int offset = 0;
+	int base_offset = 0;
+	const struct reg_cal_array *reg_off, *mod;
+	int bit_off = 0;
+	u16 value;
+	u16 value1;
+
+	int_cal = pctrl->pin_topologys->int_cal;
+	mod_sel_cal = pctrl->pin_topologys->mod_sel_cal;
+
+	if (gpio >= 0) {
+		for (i = 0; i < int_cal->size; i++)
+			if (gpio == int_cal->cal_array[i])
+				break;
+		for (j = 0; j < int_cal->reg_cal_size; j++) {
+			if (offset > i)
+				break;
+			offset += int_cal->reg[j].size;
+		}
+		reg_off = &(int_cal->reg[j-1]);
+		mod = &(mod_sel_cal->reg[j-1]);
+		bit_off = i-(offset-reg_off->size);
+		base_offset = reg_off->pmio_offset;
+
+		value = inw(pctrl->pmio_base+reg_off->pmio_offset);
+		if (mask)
+			value &= (~(1<<bit_off));
+		else
+			value |= (1<<bit_off);
+
+		outw(value, pctrl->pmio_base+reg_off->pmio_offset);
+		if (mask) {
+			value1 = readw(pctrl->pm_pmio_base+mod->pmio_offset);
+			value1 |= (1<<bit_off);
+			writew(value1, pctrl->pm_pmio_base+mod->pmio_offset);
+		} else {
+			value1 = readw(pctrl->pm_pmio_base+mod->pmio_offset);
+			value1 |= (1<<bit_off);
+			writew(value1, pctrl->pm_pmio_base+mod->pmio_offset);
+		}
+	}
+}
+
+static void zhaoxin_gpio_irq_mask(struct irq_data *d)
+{
+	zhaoxin_gpio_irq_mask_unmask(d, true);
+}
+
+static void zhaoxin_gpio_irq_unmask(struct irq_data *d)
+{
+	zhaoxin_gpio_irq_mask_unmask(d, false);
+}
+
+/*
+ * father domain irq handle
+ */
+static irqreturn_t zhaoxin_gpio_irq(int irq, void *data)
+{
+	struct zhaoxin_pinctrl *pctrl = data;
+	struct gpio_chip *gc = &pctrl->chip;
+	const struct reg_calibrate *init;
+	const struct reg_calibrate *stat_cal;
+	unsigned int i, bit_offset;
+	u16 status, enable;
+	unsigned long pending;
+	int index = 0;
+	int ret = 0;
+	int subirq;
+	unsigned int hwirq;
+
+	init = pctrl->pin_topologys->int_cal;
+	stat_cal = pctrl->pin_topologys->status_cal;
+	for (i = 0; i < init->reg_cal_size; i++) {
+		pending = 0;
+		status = readw(pctrl->pm_pmio_base + stat_cal->reg[i].pmio_offset);
+		enable = inw(pctrl->pmio_base + init->reg[i].pmio_offset);
+		enable &= status;
+		pending = enable;
+		for_each_set_bit(bit_offset, &pending, init->reg[i].size) {
+			hwirq = init->cal_array[index + bit_offset] ;
+			subirq = irq_find_mapping(gc->irq.domain, hwirq);
+			generic_handle_irq(subirq);
+		}
+
+		ret += pending ? 1 : 0;
+		index += init->reg[i].size;
+	}
+
+	return IRQ_RETVAL(ret);
+}
+
+static int zhaoxin_gpio_irq_type(struct irq_data *d, unsigned int type)
+{
+	struct gpio_chip *gc = irq_data_get_irq_chip_data(d);
+	struct zhaoxin_pinctrl *pctrl = gpiochip_get_data(gc);
+	unsigned int gpio = irqd_to_hwirq(d);
+	const struct index_cal_array *trigger_cal;
+	unsigned int pin;
+	unsigned long flags;
+	u8 index;
+	int position, point;
+	u16 value;
+	bool isup = true;
+
+	trigger_cal = pctrl->pin_topologys->trigger_cal;
+	pin = zhaoxin_gpio_to_pin(pctrl, irqd_to_hwirq(d), NULL, NULL);
+	if (type & IRQ_TYPE_EDGE_FALLING)
+		isup = true;
+	else if (type & IRQ_TYPE_EDGE_RISING)
+		isup = true;
+	else if (type & IRQ_TYPE_LEVEL_LOW)
+		isup = true;
+	else if (type & IRQ_TYPE_LEVEL_HIGH)
+		isup = false;
+
+	zhaoxin_gpio_set_gpio_mode_and_pull(pctrl, pin, isup);
+
+	for (position = 0; position < trigger_cal->size; position++)
+		if (trigger_cal->cal_array[position] == gpio)
+			break;
+
+	index = trigger_cal->index + ALIGN(position+1, 4)/4-1;
+	point = position % 4;
+
+	raw_spin_lock_irqsave(&pctrl->lock, flags);
+
+	value = zx_pad_read16(pctrl, index);
+
+	if ((type & IRQ_TYPE_EDGE_BOTH) == IRQ_TYPE_EDGE_BOTH)
+		value |= TRIGGER_BOTH_EDGE << (point*4);
+	else if (type & IRQ_TYPE_EDGE_FALLING)
+		value |= TRIGGER_FALL_EDGE << (point*4);
+	else if (type & IRQ_TYPE_EDGE_RISING)
+		value |= TRIGGER_RISE_EDGE << (point*4);
+	else if (type & IRQ_TYPE_LEVEL_LOW)
+		value |= TRIGGER_LOW_LEVEL << (point*4);
+	else if (type & IRQ_TYPE_LEVEL_HIGH)
+		value |= TRIGGER_HIGH_LEVEL << (point*4);
+	else
+		pr_debug(pctrl, "%s wrang type\n", __func__);
+
+	zx_pad_write16(pctrl, index, value);
+
+	if (type & IRQ_TYPE_EDGE_BOTH)
+		irq_set_handler_locked(d, handle_edge_irq);
+	else if (type & IRQ_TYPE_LEVEL_MASK)
+		irq_set_handler_locked(d, handle_level_irq);
+	raw_spin_unlock_irqrestore(&pctrl->lock, flags);
+
+	return 0;
+}
+
+static int zhaoxin_gpio_irq_wake(struct irq_data *d, unsigned int on)
+{
+	struct gpio_chip *gc = irq_data_get_irq_chip_data(d);
+	struct zhaoxin_pinctrl *pctrl = gpiochip_get_data(gc);
+	unsigned int pin;
+
+	pin = zhaoxin_gpio_to_pin(pctrl, irqd_to_hwirq(d), NULL, NULL);
+	if (pin) {
+		if (on)
+			enable_irq_wake(pctrl->irq);
+		else
+			disable_irq_wake(pctrl->irq);
+	}
+
+	return 0;
+}
+
+static int zhaoxin_gpio_add_pin_ranges(struct gpio_chip *gc)
+{
+	struct zhaoxin_pinctrl *pctrl = gpiochip_get_data(gc);
+	int ret, i;
+
+	for (i = 0; i < pctrl->pin_map_size; i++) {
+		struct zhaoxin_pin_map2_gpio *map = &pctrl->pin_maps[i];
+
+		if (map->zhaoxin_range_gpio_base == ZHAOXIN_GPIO_BASE_NOMAP)
+			continue;
+		ret = gpiochip_add_pin_range(&pctrl->chip, dev_name(pctrl->dev),
+					     map->zhaoxin_range_gpio_base,
+					     map->zhaoxin_range_pin_base,
+					     map->zhaoxin_range_pin_size);
+		if (ret) {
+			dev_err(pctrl->dev, "failed to add GPIO pin range\n");
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static unsigned int zhaoxin_gpio_ngpio(const struct zhaoxin_pinctrl *pctrl)
+{
+	const struct zhaoxin_pin_map2_gpio *pin_maps;
+	unsigned int ngpio = 0;
+	int i;
+
+	for (i = 0; i < pctrl->pin_map_size; i++) {
+		pin_maps = &pctrl->pin_maps[i];
+		if (pin_maps->zhaoxin_range_gpio_base == ZHAOXIN_GPIO_BASE_NOMAP)
+			continue;
+		if (pin_maps->zhaoxin_range_gpio_base + pin_maps->zhaoxin_range_pin_size > ngpio)
+			ngpio = pin_maps->zhaoxin_range_gpio_base +
+				pin_maps->zhaoxin_range_pin_size;
+	}
+
+	return ngpio;
+}
+
+static int zhaoxin_gpio_probe(struct zhaoxin_pinctrl *pctrl, int irq)
+{
+	int ret;
+	struct gpio_irq_chip *girq;
+
+	pctrl->chip = zhaoxin_gpio_chip;
+
+	pctrl->chip.ngpio = zhaoxin_gpio_ngpio(pctrl);
+	pctrl->chip.label = dev_name(pctrl->dev);
+	pctrl->chip.parent = pctrl->dev;
+	pctrl->chip.base = -1;
+	pctrl->chip.add_pin_ranges = zhaoxin_gpio_add_pin_ranges;
+
+	pctrl->irq = irq;
+
+	pctrl->irqchip.name = dev_name(pctrl->dev);
+	pctrl->irqchip.irq_ack = zhaoxin_gpio_irq_ack;
+	pctrl->irqchip.irq_mask = zhaoxin_gpio_irq_mask;
+	pctrl->irqchip.irq_unmask = zhaoxin_gpio_irq_unmask;
+	pctrl->irqchip.irq_set_type = zhaoxin_gpio_irq_type;
+	pctrl->irqchip.irq_set_wake = zhaoxin_gpio_irq_wake;
+	pctrl->irqchip.flags = IRQCHIP_MASK_ON_SUSPEND;
+	/*
+	 * father domain irq
+	 */
+	ret = devm_request_irq(pctrl->dev, irq, zhaoxin_gpio_irq, IRQF_SHARED | IRQF_NO_THREAD,
+			       dev_name(pctrl->dev), pctrl);
+	if (ret) {
+		dev_err(pctrl->dev, "failed to request interrupt\n");
+		return ret;
+	}
+	girq = &pctrl->chip.irq;
+	girq->chip = &pctrl->irqchip;
+	/* This will let us handle the IRQ in the driver */
+	girq->parent_handler = NULL;
+	girq->num_parents = 0;
+	girq->default_type = IRQ_TYPE_NONE;
+	girq->handler = handle_bad_irq;
+	ret = devm_gpiochip_add_data(pctrl->dev, &pctrl->chip, pctrl);
+	if (ret) {
+		dev_err(pctrl->dev, "failed to register gpiochip\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int zhaoxin_pinctrl_pm_init(struct zhaoxin_pinctrl *pctrl)
+{
+	return 0;
+}
+
+static int zhaoxin_pinctrl_probe(struct platform_device *pdev,
+				 const struct zhaoxin_pinctrl_soc_data *soc_data)
+{
+	struct zhaoxin_pinctrl *pctrl;
+	int  ret, i, irq;
+	struct resource *res;
+	void __iomem *regs;
+
+	pctrl = devm_kzalloc(&pdev->dev, sizeof(*pctrl), GFP_KERNEL);
+	if (!pctrl)
+		return -ENOMEM;
+	pctrl->dev = &pdev->dev;
+	pctrl->soc = soc_data;
+	raw_spin_lock_init(&pctrl->lock);
+	pctrl->pin_topologys = pctrl->soc->pin_topologys;
+	pctrl->pin_map_size = pctrl->soc->pin_map_size;
+	pctrl->pin_maps = devm_kcalloc(&pdev->dev, pctrl->pin_map_size, sizeof(*pctrl->pin_maps),
+				       GFP_KERNEL);
+	if (!pctrl->pin_maps)
+		return -ENOMEM;
+	for (i = 0; i < pctrl->pin_map_size; i++) {
+		struct zhaoxin_pin_map2_gpio *community = &pctrl->pin_maps[i];
+		*community = pctrl->soc->zhaoxin_pin_maps[i];
+	}
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	regs = devm_ioremap_resource(&pdev->dev, res);
+	if (IS_ERR(regs))
+		return PTR_ERR(regs);
+
+	pctrl->pm_pmio_base = regs;
+	pctrl->pmio_base = 0x800;
+	pctrl->pmio_rx90 = 0x90;
+	pctrl->pmio_rx8c = 0x8c;
+	irq = platform_get_irq(pdev, 0);
+	if (irq < 0)
+		return irq;
+
+	ret = zhaoxin_pinctrl_pm_init(pctrl);
+	if (ret)
+		return ret;
+	pctrl->pctldesc = zhaoxin_pinctrl_desc;
+	pctrl->pctldesc.name = dev_name(&pdev->dev);
+	pctrl->pctldesc.pins = pctrl->soc->pins;
+	pctrl->pctldesc.npins = pctrl->soc->npins;
+	pctrl->pctldev = devm_pinctrl_register(&pdev->dev, &pctrl->pctldesc, pctrl);
+	if (IS_ERR(pctrl->pctldev)) {
+		dev_err(&pdev->dev, "failed to register pinctrl driver\n");
+		return PTR_ERR(pctrl->pctldev);
+	}
+	ret = zhaoxin_gpio_probe(pctrl, irq);
+
+	if (ret)
+		return ret;
+	platform_set_drvdata(pdev, pctrl);
+	return 0;
+}
+
+int zhaoxin_pinctrl_probe_by_hid(struct platform_device *pdev)
+{
+	const struct zhaoxin_pinctrl_soc_data *data;
+
+	data = device_get_match_data(&pdev->dev);
+	if (!data)
+		return -ENODATA;
+
+	return zhaoxin_pinctrl_probe(pdev, data);
+}
+EXPORT_SYMBOL_GPL(zhaoxin_pinctrl_probe_by_hid);
+
+int zhaoxin_pinctrl_probe_by_uid(struct platform_device *pdev)
+{
+	const struct zhaoxin_pinctrl_soc_data *data;
+
+	data = zhaoxin_pinctrl_get_soc_data(pdev);
+	if (IS_ERR(data))
+		return PTR_ERR(data);
+
+	return zhaoxin_pinctrl_probe(pdev, data);
+}
+EXPORT_SYMBOL_GPL(zhaoxin_pinctrl_probe_by_uid);
+
+
+const struct zhaoxin_pinctrl_soc_data *zhaoxin_pinctrl_get_soc_data(struct platform_device *pdev)
+{
+	const struct zhaoxin_pinctrl_soc_data *data = NULL;
+	const struct zhaoxin_pinctrl_soc_data **table;
+	struct acpi_device *adev;
+	unsigned int i;
+
+	adev = ACPI_COMPANION(&pdev->dev);
+	if (adev) {
+		const void *match = device_get_match_data(&pdev->dev);
+
+		table = (const struct zhaoxin_pinctrl_soc_data **)match;
+		for (i = 0; table[i]; i++) {
+			if (!strcmp(adev->pnp.unique_id, table[i]->uid)) {
+				data = table[i];
+				break;
+			}
+		}
+	} else {
+		const struct platform_device_id *id;
+
+		id = platform_get_device_id(pdev);
+		if (!id)
+			return ERR_PTR(-ENODEV);
+
+		table = (const struct zhaoxin_pinctrl_soc_data **)id->driver_data;
+		data = table[pdev->id];
+	}
+
+	return data ?: ERR_PTR(-ENODATA);
+}
+EXPORT_SYMBOL_GPL(zhaoxin_pinctrl_get_soc_data);
+
+#ifdef CONFIG_PM_SLEEP
+int zhaoxin_pinctrl_suspend_noirq(struct device *dev)
+{
+	return 0;
+}
+EXPORT_SYMBOL_GPL(zhaoxin_pinctrl_suspend_noirq);
+
+int zhaoxin_pinctrl_resume_noirq(struct device *dev)
+{
+	return 0;
+}
+EXPORT_SYMBOL_GPL(zhaoxin_pinctrl_resume_noirq);
+#endif /* CONFIG_PM_SLEEP */
+
+MODULE_AUTHOR("www.zhaoxin.com");
+MODULE_DESCRIPTION("Shanghai Zhaoxin pinctrl driver");
+MODULE_VERSION(DRIVER_VERSION);
+MODULE_LICENSE("GPL v2");
diff --git a/drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.h b/drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.h
new file mode 100644
index 000000000000..88ef00796729
--- /dev/null
+++ b/drivers/pinctrl/zhaoxin/pinctrl-zhaoxin.h
@@ -0,0 +1,135 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
+/*
+ * zhaoxin pinctrl common code
+ * Copyright(c) 2023 Shanghai Zhaoxin Corporation. All rights reserved.
+ */
+
+#ifndef PINCTRL_zhaoxin_H
+#define PINCTRL_zhaoxin_H
+
+#include <linux/bits.h>
+#include <linux/compiler_types.h>
+#include <linux/gpio/driver.h>
+#include <linux/irq.h>
+#include <linux/kernel.h>
+#include <linux/pm.h>
+#include <linux/pinctrl/pinctrl.h>
+#include <linux/spinlock_types.h>
+
+struct platform_device;
+struct device;
+
+/*
+ * struct zhaoxin_pingroup pin define
+ */
+struct zhaoxin_pingroup {
+	const char *name;
+	const unsigned int *pins;
+	size_t npins;
+	unsigned short mode;
+	const unsigned int *modes;
+};
+
+/*
+ * struct zhaoxin_function
+ */
+struct zhaoxin_function {
+	const char *name;
+	const char * const *groups;
+	size_t ngroups;
+};
+
+/*
+ * struct zhaoxin_pin_map2_gpio
+ * @zhaoxin_range_pin_base
+ * @size: pin number
+ * @zhaoxin_range_gpio_base
+ */
+struct zhaoxin_pin_map2_gpio {
+	unsigned int zhaoxin_range_pin_base;
+	unsigned int zhaoxin_range_pin_size;
+	int zhaoxin_range_gpio_base;
+};
+
+#define MAX_GPIO		256
+
+struct reg_cal_array {
+	int pmio_offset;
+	int size;
+};
+
+struct reg_calibrate {
+	const struct reg_cal_array *reg;
+	const int reg_cal_size;
+	const int *cal_array;
+	const int size;
+};
+
+struct index_cal_array {
+	int reg_port_base;
+	int reg_data_base;
+	int index;
+	int *cal_array;
+	int size;
+};
+
+struct zhaoxin_pin_topology {
+	const struct reg_calibrate *int_cal;
+	const struct reg_calibrate *mod_sel_cal;
+	const struct reg_calibrate *status_cal;
+	const struct index_cal_array *gpio_in_cal;
+	const struct index_cal_array *gpio_out_cal;
+	const struct index_cal_array *gpio_dir_cal;
+	const struct index_cal_array *trigger_cal;
+};
+
+#define TRIGGER_FALL_EDGE	0
+#define TRIGGER_RISE_EDGE	1
+#define TRIGGER_BOTH_EDGE	2
+#define TRIGGER_LOW_LEVEL	3
+#define TRIGGER_HIGH_LEVEL	4
+
+#define	ZHAOXIN_GPIO_BASE_NOMAP -1
+
+struct zhaoxin_pinctrl_soc_data {
+	const char *uid;
+	const struct pinctrl_pin_desc *pins;
+	size_t npins;
+	const struct zhaoxin_pingroup *groups;
+	size_t ngroups;
+	const struct zhaoxin_function *functions;
+	size_t nfunctions;
+	const struct zhaoxin_pin_topology *pin_topologys;
+	const struct zhaoxin_pin_map2_gpio *zhaoxin_pin_maps;
+	size_t pin_map_size;
+};
+
+const struct zhaoxin_pinctrl_soc_data *zhaoxin_pinctrl_get_soc_data(struct platform_device *pdev);
+
+struct zhaoxin_pinctrl {
+	struct device *dev;
+	raw_spinlock_t lock;
+	struct pinctrl_desc pctldesc;
+	struct pinctrl_dev *pctldev;
+	struct gpio_chip chip;
+	struct irq_chip irqchip;
+	const struct zhaoxin_pinctrl_soc_data *soc;
+	const struct zhaoxin_pin_topology *pin_topologys;
+	struct zhaoxin_pin_map2_gpio *pin_maps;
+	size_t pin_map_size;
+	int irq;
+	int pmio_base;
+	void __iomem *pm_pmio_base;
+	int pmio_rx90;
+	int pmio_rx8c;
+};
+
+int zhaoxin_pinctrl_probe_by_hid(struct platform_device *pdev);
+int zhaoxin_pinctrl_probe_by_uid(struct platform_device *pdev);
+
+#ifdef CONFIG_PM_SLEEP
+int zhaoxin_pinctrl_suspend_noirq(struct device *dev);
+int zhaoxin_pinctrl_resume_noirq(struct device *dev);
+#endif
+
+#endif /* PINCTRL_zhaoxin_H */
-- 
2.20.1

